Automattic/newspack-popups
# refactor: segmentation data

### All Submissions:






### Changes proposed in this Pull Request:



This PR implements a significant refactor of segmentation data storage and retrieval for improved performance. It does the following:

* Creates two new tables with a column and index structure optimized for performance:
  - The "readers" table replaces the legacy "transients" table and normalizes data that was previously serialized in a single column into separate, indexable columns
  - The "reader events" table replaces the legacy "events" table and stores info on not just `post_read` events, but also non-article page views, prompt views, donations, and newsletter subscriptions. In the future, when we're able to support a JS-based alternative to `amp-access`, this table may be offloaded from the server to `localStorage`.
  - Because the above are stored as events, we no longer replicate the data in the readers table, except for keeping a count of read categories, article views, and non-article page views. A reader is now considered a donor if they have at least one "donation" event associated with their unique client ID, for example.
  - The legacy readers table is retained, but existing readers who visit the site will be migrated to the new model and their legacy data deleted. Migrated data includes a lifetime count of posts and categories read, donations, and subscriptions. Prompt views and dismissals will not be migrated as we no longer need to store permanent suppression data. Old data lacking donations and subscriptions will be purged after 30 days.
  - The legacy events table can be dropped if it exists. (**Note:** I didn't implement this in case we need to roll back the refactor. We can push a new PR to drop these tables once we confirm the refactor is successful.) Since `post_read` events are no longer needed after 1 hour, this table can be safely deprecated. It will result in a 1-hour period where all returning readers are treated as not having seen any prompts, after which the new data model will start tracking prompt views as usual. I doubt this will be noticeable to most readers.
  - The previous legacy handling (v1 transients data) is deprecated, since no existing Newspack sites should have any data in the original model at this point.
* Improved data purging logic: we purge article and page view events every hour (as opposed to every 30 days) as we only need to keep a record of unique post IDs and timestamps during a single session. Prompt view events are kept for 30 days (to facilitate upcoming new frequency options for daily/weekly/monthly periods), and donation/subscription events are kept permanently.

The above changes should give us some significant performance boosts:

* We avoid the use of `LIKE` queries and instead base all queries on indexed columns
* We avoid unlimited `DELETE` queries and instead chunk rows to delete into groups of 1000, to avoid replication lag
* We avoid potentially unlimited row lengths caused by the current serialized data model
* We should hopefully limit the growth of the events table by purging article/page view events more often, although this may be offset by moving prompt views/donations/subscriptions to this table
* We still rely on the persistent WP cache (if available) for retrieving reader and events data per page view. Events are logged to a flat file and parsed to the DB every 15 minutes. This behavior is unchanged from the current.

The refactor should result in few if any reader-facing changes, except for faster site performance at scale. The exceptions:

* A "session" is now considered one whole hour instead of 45 minutes. This is simply because WP cron supports the `hourly` period, but it shouldn't make a huge difference in segmentation behavior. 45 minutes was an arbitrary choice to begin with.
* `post_read` events are renamed as `article_view` for clarity. The behavior is the same—a view of a singular post of any public-facing post type is considered an article view.
* Non-article page views (archives, search results) are now tracked as `page_view`. This will become important for upcoming prompt frequency options based on total page views.
* Article/page views are now only de-deduplicated for the current session, instead of for the lifetime of the reader. For example, if you visit or refresh the same post multiple times within an hour, that will still count as only a single article view. If you visit that same article again the next day, it will count as a new article view. This is closer to how Google Analytics counts page views per session, and will be important for upcoming prompt frequency options based on page views.
* We no longer store subscription/donation data on the reader data itself, instead storing these as separate events that can be cross-queried using the client ID. In the future, if we move the events data to `localStorage`, we may want to retain subscription and donation data on the server so we don't lose that data if the browser cache is purged.

### How to test the changes in this Pull Request:

It's hard to quantify the performance improvements that this refactor should enable without running the plugin at scale, so we should plan to test this refactor with a small number of sites before rolling out completely. Other than performance issues:

* This refactor should result in **no reader-facing changes**. So please thoroughly test all important prompt and segmentation features as a reader and ensure that nothing unexpected happens.
* Also test the editor experience, particularly the Campaigns Analytics dashboard, and make sure you see nothing unexpected there, too.
* Due diligence: test the data migration handling by running the plugin on `master` first. 
  1. In a new incognito window, visit several posts with categories, subscribe via a newsletter signup prompt, and donate via a donate prompt. Note which posts/categories you viewed so you can confirm those views are migrated to the new model.
  2. inspect your DB tables directly via Adminer or Sequel Pro and note the client ID for your reader activity as created in the `wp_newspack_campaigns_transients` table.
  3. Switch to this branch and **in the same incognito window** (so that your client ID remains the same) continue navigating the site and confirm that you still see the prompts you'd expect to see based on your segment.
  4. Inspect your DB tables and confirm that a row is created for your client ID in the `wp_newspack_campaigns_readers` table that carries over the previously viewed posts and categories. Also confirm that events were created for your client ID in the `wp_newspack_campaigns_events` table for your subscription and donation.
* Also make sure to test in both AMP and non-AMP mode.

### Other information:
